core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java

```diff
@@ -139,9 +166,14 @@ private void reroute() {
             if (lifecycle.stopped()) {
                 return;
             }
+            if (rerouting.compareAndSet(false, true) == false) {
```

Comment:
I think it would be clearer if we acquire a lock here somehow using `tryLock()`  this would make it clear that there is only one reroute happening at the same time?